File: test/code-snippets/java.java
package com.sunlightjs.awesome;

import java.text.*;
import java.io.BufferedStreamReader;
import java.lang.annotation.*;
import java.lang.StackTraceElement;

public class MyClass extends ExtendedClass implements Interfacable1, Interfacable2, Interfacable3 {
    private int foo;

    public MyClass(int parameter1) {
        super();
        
        this.foo = (CastingTest1)foo;
        this.foo = ((CastingTest2)this).doStuff();
        if (notACast) methodCall();
        if (this.foo instanceof java.lang.CharSequence) {
            throw new Exception();
        }
    }
    
    /**
     * How about an inner class?
     */
    private class InnerEvenIterator {
        //start stepping through the array from the beginning
        private int next = 0;
        
        public boolean hasNext() {
            //check if a current element is the last in the array
            return (next <= SIZE - 1);
        }
        
        public int getNext() {
            //record a value of an even index of the array
            int retValue = arrayOfInts[next];
            //get the next even element
            next += 2;
            return retValue;
        }
    }
    
    private void anonymousInnerClassTest() throws CheckedException1, CheckedException2, CheckedException3 {
        new Thread(new Runnable() {
          public void run() {
            try {
              while (true) {
                sleep(1000); System.out.print(".");
              }
            }
            catch(InterruptedException ex) {}
          }
        }).start();
        
        //http://www.roseindia.net/javatutorials/anonymous_innerclassestutorial.shtml
        Vector myVector = new Vector(3) {{
            add("Heinz"); add("John"); add("Anton");
        }};
    }
    
    /*
    multi line
    comment
    */
    protected abstract void abstractMethod(Kissable lips, char c, ArrayParameter[] arrayParameter);
    
    /**
     * Stolen mostly from http://download.oracle.com/javase/tutorial/java/generics/bounded.html
     * because I haven't written Java since I was an idiot in college
     */
    @Documented @AnnotationWithArguments(name = "this \"is\" a string\\")
    @Override public static <T extends GenericExtended & GenericImplemented<? super T>> T genericMethod(GenericParam<? extends T> genericParam) {
        //wait, so Java seriously uses the ampersand for interfaces in generics?
        
        MyCustomMap1<char, AnotherGeneric> map = new MyCustomMap2<FirstGeneric, SecondGeneric>();
        fully1.qualified1.ClassName1 fqc = new fully2.qualified2.ClassName2();
        
        int length = 2;
        ArrayTest1[] arrayOStuff = length > 0 ? new ArrayTest2[length] : new ArrayTest3[12];
        
        return true ? false : null;
    }
}

//enum test: http://download.oracle.com/javase/1.5.0/docs/guide/language/enums.html
public enum Planet {
    MERCURY (3.303e+23, 2.4397e6),
    VENUS   (4.869e+24, 6.0518e6),
    EARTH   (5.976e+24, 6.37814e6),
    MARS    (6.421e+23, 3.3972e6),
    JUPITER (1.9e+27,   7.1492e7),
    SATURN  (5.688e+26, 6.0268e7),
    URANUS  (8.686e+25, 2.5559e7),
    NEPTUNE (1.024e+26, 2.4746e7),
    PLUTO   (1.27e+22,  1.137e6);

    private final double mass;   // in kilograms
    private final double radius; // in meters
    Planet(double mass, double radius) {
        this.mass = mass;
        this.radius = radius;
    }
    public double mass()   { return mass; }
    public double radius() { return radius; }

    // universal gravitational constant  (m3 kg-1 s-2)
    public static final double G = 6.67300E-11;

    public double surfaceGravity() {
        return G * mass / (radius * radius);
    }
    public double surfaceWeight(double otherMass) {
        return otherMass * surfaceGravity();
    }
}

File: test/java.js
// sunlight-x: Intelligent Syntax Highlighting, Modernized
// Copyright 2017 Leung Wing-chung. All rights reserved.
// Use of this source code is governed by a Apache License Version 2.0, that can
// be found in the LICENSE file.

// @flow
import { TestSupportForFile } from "./fixtures/testsupport.js";

let testSupport: TestSupportForFile;
describe("Javascript tests", function() {
  beforeAll(function() {
    testSupport = new TestSupportForFile("java.java", "java");
  });
  it("extends keyword", function() {
    testSupport.AssertContentExists("keyword", "extends");
  });
  it("implements keyword", function() {
    testSupport.AssertContentExists("keyword", "implements");
  });
  it("package keyword", function() {
    testSupport.AssertContentExists("keyword", "package");
  });
  it("import keyword", function() {
    testSupport.AssertContentExists("keyword", "import");
  });
  it("private keyword", function() {
    testSupport.AssertContentExists("keyword", "private");
  });
  it("int keyword", function() {
    testSupport.AssertContentExists("keyword", "int");
  });
  it("public keyword", function() {
    testSupport.AssertContentExists("keyword", "public");
  });
  it("super keyword", function() {
    testSupport.AssertContentExists("keyword", "super");
  });
  it("protected keyword", function() {
    testSupport.AssertContentExists("keyword", "protected");
  });
  it("abstract keyword", function() {
    testSupport.AssertContentExists("keyword", "abstract");
  });
  it("char keyword", function() {
    testSupport.AssertContentExists("keyword", "char");
  });
  it("void keyword", function() {
    testSupport.AssertContentExists("keyword", "void");
  });
  it("null keyword", function() {
    testSupport.AssertContentExists("keyword", "null");
  });
  it("this keyword", function() {
    testSupport.AssertContentExists("keyword", "this");
  });
  it("false keyword", function() {
    testSupport.AssertContentExists("keyword", "false");
  });
  it("true keyword", function() {
    testSupport.AssertContentExists("keyword", "true");
  });
  it("true keyword", function() {
    testSupport.AssertContentExists("keyword", "enum");
  });
  it("while keyword", function() {
    testSupport.AssertContentExists("keyword", "while");
  });
  it("throws keyword", function() {
    testSupport.AssertContentExists("keyword", "throws");
  });

  it("cast", function() {
    testSupport.AssertContentExists("named-ident", "CastingTest1");
  });
  it("cast", function() {
    testSupport.AssertContentExists("named-ident", "CastingTest2");
  });
  it("class name", function() {
    testSupport.AssertContentExists("named-ident", "MyClass");
  });
  it("class name", function() {
    testSupport.AssertContentExists("named-ident", "ExtendedClass");
  });
  it("interface name", function() {
    testSupport.AssertContentExists("named-ident", "Interfacable1");
  });
  it("interface name", function() {
    testSupport.AssertContentExists("named-ident", "Interfacable2");
  });
  it("interface name", function() {
    testSupport.AssertContentExists("named-ident", "Interfacable3");
  });
  it("array parameter", function() {
    testSupport.AssertContentExists("named-ident", "ArrayParameter");
  });
  it("type constraint for generic", function() {
    testSupport.AssertContentExists("named-ident", "GenericExtended");
  });
  it("type constraint for generic after &", function() {
    testSupport.AssertContentExists("named-ident", "GenericImplemented");
  });
  it("generic parameter", function() {
    testSupport.AssertContentExists("named-ident", "GenericParam");
  });
  it("fully qualified class name", function() {
    testSupport.AssertContentExists("named-ident", "ClassName1");
  });
  it("fully qualified class name after new", function() {
    testSupport.AssertContentExists("named-ident", "ClassName2");
  });
  it("generic parameter", function() {
    testSupport.AssertContentExists("named-ident", "AnotherGeneric");
  });
  it("generic parameter", function() {
    testSupport.AssertContentExists("named-ident", "FirstGeneric");
  });
  it("generic parameter", function() {
    testSupport.AssertContentExists("named-ident", "SecondGeneric");
  });
  it("generic class name", function() {
    testSupport.AssertContentExists("named-ident", "MyCustomMap1");
  });
  it("generic class name", function() {
    testSupport.AssertContentExists("named-ident", "MyCustomMap2");
  });
  it("array declaration", function() {
    testSupport.AssertContentExists("named-ident", "ArrayTest1");
  });
  it("array instantiation with variable", function() {
    testSupport.AssertContentExists("named-ident", "ArrayTest2");
  });
  it("array instantiation with integer", function() {
    testSupport.AssertContentExists("named-ident", "ArrayTest3");
  });
  it("enum name", function() {
    testSupport.AssertContentExists("named-ident", "Planet");
  });
  it("caught exception name", function() {
    testSupport.AssertContentExists("named-ident", "InterruptedException");
  });
  it("checked exception 1", function() {
    testSupport.AssertContentExists("named-ident", "CheckedException1");
  });
  it("checked exception 2", function() {
    testSupport.AssertContentExists("named-ident", "CheckedException2");
  });
  it("checked exception 3", function() {
    testSupport.AssertContentExists("named-ident", "CheckedException3");
  });
  it("fully qualified type name after instanceof", function() {
    testSupport.AssertContentExists("named-ident", "CharSequence");
  });
  it("fully qualified type name after import", function() {
    testSupport.AssertContentExists("named-ident", "StackTraceElement");
  });

  it("method parameter", function() {
    testSupport.AssertContentExists("ident", "parameter1");
  });
  it("method parameter", function() {
    testSupport.AssertContentExists("ident", "arrayParameter");
  });
  it("abstract method name", function() {
    testSupport.AssertContentExists("ident", "abstractMethod");
  });
  it("generic method name", function() {
    testSupport.AssertContentExists("ident", "genericMethod");
  });
  it("generic parameter", function() {
    testSupport.AssertContentExists("ident", "genericParam");
  });
  it("fully qualified package name", function() {
    testSupport.AssertContentExists("ident", "fully1");
  });
  it("fully qualified package name", function() {
    testSupport.AssertContentExists("ident", "fully2");
  });
  it("fully qualified package name", function() {
    testSupport.AssertContentExists("ident", "qualified1");
  });
  it("fully qualified package name", function() {
    testSupport.AssertContentExists("ident", "qualified2");
  });
  it("variable name", function() {
    testSupport.AssertContentExists("ident", "myVector");
  });
  it("if (notACast) methodCall()", function() {
    testSupport.AssertContentExists("ident", "notACast");
  });

  it("Documented annotation", function() {
    testSupport.AssertContentExists("annotation", "@Documented");
  });
  it("AnnotationWithArguments annotation", function() {
    testSupport.AssertContentExists("annotation", "@AnnotationWithArguments");
  });
  it("annotation", function() {
    testSupport.AssertContentExists("annotation", "@Override");
  });

  it("single line comment", function() {
    testSupport.AssertContentExists(
      "comment",
      "//wait, so Java seriously uses the ampersand for interfaces in generics?"
    );
  });
  it("multi line comment", function() {
    testSupport.AssertContentExists(
      "comment",
      "/*\n\tmulti line\n\tcomment\n\t*/"
    );
  });

  it("string with escapes", function() {
    testSupport.AssertContentExists("string", '"this \\"is\\" a string\\\\"');
  });
});
